ffi/src/lib.rs

```diff
@@ -25,3 +35,150 @@ pub enum FfiResult {
     /// FFI function executed successfully
     Ok = 0_i32,
 }
+
+/// Implement [`Handle`] for given types with first argument as the initial handle id.
+#[macro_export]
+macro_rules! handles {
+    ( $id:expr, $ty:ty $(, $other:ty)* $(,)? ) => {
+        impl $crate::Handle for $ty {
+            const ID: $crate::HandleId = $id;
+        }
+
+        $crate::handles! {$id + 1, $( $other, )*}
+    };
+    ( $id:expr, $(,)? ) => {};
+}
+
+/// Generate FFI equivalent implementation of the requested trait method (e.g. Clone, Eq, Ord)
+#[macro_export]
+macro_rules! gen_ffi_impl {
+    (@null_check_stmts $( $ptr:ident ),+ ) => {
+    $(  if $ptr.is_null() {
+            return $crate::FfiResult::ArgIsNull;
+        } )+
+    };
+    ( Clone: $( $other:ty ),+ $(,)? ) => {
+        /// FFI function equivalent of [`Clone::clone`]
+        ///
+        /// # Safety
+        ///
+        /// All of the given pointers must be valid and the given handle id must match the expected
+        /// pointer type
+        #[no_mangle]
+        pub unsafe extern "C" fn __clone(
+            handle_id: $crate::HandleId,
+            handle_ptr: *const core::ffi::c_void,
+            output_ptr: *mut *mut core::ffi::c_void
+        ) -> $crate::FfiResult {
+            gen_ffi_impl!{@null_check_stmts handle_ptr, output_ptr}
+
+            // NOTE: Unreachable pattern occurs if at least two handle ids resolve to
+            // the same value. Denying it reduces chance of such implementation error
+            #[deny(unreachable_patterns)]
```

Comment:
I don't understand why this lint doesn't activate when I'm compiling Iroha. I can implement this check manually, but this seemed really convenient and short